### PR TITLE
Use file to detect the mime-type

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
RONSEAL

### CODE REVIEW

**Review by benarmston — COMMENTED** [review]
I've rebased this on the latest `master`.

This fixes the issue.  The result files for bowtie example script are now correctly detected as either `application/octect-stream` or `text/plain`.

I've added a question about how the mime type is being determined.

File: api/app/models/job_file.rb [hunk 1]
```diff
@@ -176,6 +176,18 @@ def payload
     @payload = File.read path
   end
 
+  def mime_type
+    # Protect against bad/missing paths, this shouldn't happen in practice
+    return nil unless exists?
+    cmd = ['file', '--mime-type', path]
+    results = FlightJobScriptAPI::Subprocess.default.run(cmd, '')
```
[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/app/models/job_file.rb:183]: I had a look at using a `ruby` library, but couldn't find one I particularly liked. So I wrapped `file`  instead.

In regard to `mimemagic`, it looks like it uses the `freedesktop.org.xml` but does the magic number detection itself. This isn't equivalent to how `file` works.

---

Instead something that integrates with `libmagic` is probably preferable:

* [ruby-magic](https://github.com/kwilczynski/ruby-magic)
* [ffi-magic](https://github.com/manveru/ffi-magic)

`ruby-magic` has the nicer interface, but monkey patches the `File` class :frowning_face:. The `ffi-magic` more closely wraps `libmagic`, so its interface is a bit more verbose.

[the remark this answers is not in the document]
Comment (benarmston) [reply api/app/models/job_file.rb:183]: Why would using `libmagic` be preferable to using freedesktop.org's shared mime database?  The freedesktop.org shared mime database is the newer of the two and is the one that is used by Linux desktop applications such as file managers.

What do you mean by  `mimemagic` not being equivalent to how `file` works?  `file` uses an algorithm implemented in `libmagic` to determine the content type of a file from its own database.  `mimemagic` implements the algorithm specified by freedesktop.org to determine the content type from the database provided by freedesktop.org.  These are broadly the same.

The results could differ if a custom file type is registered in only one of the two locations.  So the questions becomes, should we be consistent with:

1. The `file` CLI and applications using `libmagic`
2. Desktop software such as KDE, Gnome, Xfce etc..

[the remark this answers is not in the document]
Comment (WilliamMcCumstie) [reply api/app/models/job_file.rb:183]: On review, I agree with `libmagic`  and `freedesktop` being broadly the same. However, I still have a preference for `libmagic`.

> The freedesktop.org shared mime database is the newer of the two and is the one that is used by Linux desktop applications such as file managers.

That is certainly an advantage of `freedesktop`, however `libmagic`/`file` is still widely in use.

In regards to the "file manager" applications, there is an added complexity with `flight-file-manager`. It has recently gained the ability to preform mime type detection with `libmagic`:

https://github.com/openflighthpc/flight-file-manager/pull/20

Looking through `npm`, I couldn't see a suitable library which implements the `freedesktop` algorithm.

> `mimemagic` implements the algorithm specified by freedesktop.org to determine the content type from the database provided by freedesktop.org. These are broadly the same.

The `freedesktop` database isn't the concern, it is the ruby implementation of the algorithm. `mimemagic` is probably fine, I had a quick look over the code and it looks to be in order. However using `libmagic` resolves this concern.

> The results could differ if a custom file type is registered in only one of the two locations.

That is a natural consequence of `linux` using two data sources for mime type detection. In the above case, the desktop applications (`gnome`/`kde`) would disagree with the `file` utility.

> So the questions becomes, should we be consistent with ....

It is more important that the `flight-*-suite` is consistent within itself. On this count, `libmagic` wins out as it has bindings in both `node` and `ruby`.

AFAICS, the `freedesktop` algorithm has only been implemented within `ruby`, I couldn't see a suitable `node` package. The closest options where:


* [system-mime](https://www.npmjs.com/package/system-mime?activeTab=readme) - Doesn't support the magic, or
* [mimetype-description](https://github.com/myhau/mimetype-descriptions) - Doesn't implement the algorithm

**Comment by WilliamMcCumstie** [comment]
Rebased on master

**Comment by WilliamMcCumstie** [comment]
Switched to using `ffi-magic`

**Comment by benarmston** [comment]
Rebased on `master`.

**Comment by benarmston** [comment]
@WilliamMcCumstie I think that you are correct in your assessment of freedesktop mime type database and npm.  Using `libmagic` is probably the better option.

I've added a fix to the mime type detection to return `text/plain` instead of `text/plain; charset=whatever`.  With that done this LGTM.

**Pull request merged**